tasty-plutus/test/Properties/Main.hs

```diff
@@ -0,0 +1,172 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Main (main) where
+
+--------------------------------------------------------------------------------
+
+import Prelude hiding (fmap, ($), (&&), (+), (-), (==))
+
+import Ledger.Crypto (PubKeyHash)
+import Plutus.V1.Ledger.Value (Value)
+import Test.QuickCheck.Plutus.Instances ()
+import Test.Tasty (TestTree, defaultMain, localOption)
+import Test.Tasty.Plutus.Context (
+  ContextBuilder,
+  Purpose (ForSpending),
+  paysToPubKey,
+  paysToSelf,
+ )
+import Test.Tasty.Plutus.Script.Property (scriptProperty)
+import Test.Tasty.Plutus.TestData (
+  Generator (GenForSpending),
+  Methodology (Methodology),
+  Outcome (Fail, Pass),
+  TestItems (
+    ItemsForSpending,
+    spendCB,
+    spendDatum,
+    spendOutcome,
+    spendRedeemer,
+    spendValue
+  ),
+ )
+import Test.Tasty.Plutus.WithScript (toTestValidator, withValidator)
+import Test.Tasty.QuickCheck (Gen, QuickCheckTests (QuickCheckTests), arbitrary, genericShrink, oneof)
+
+--------------------------------------------------------------------------------
+
+import Plutus.V1.Ledger.Contexts (ScriptContext, TxOut)
+import Plutus.V1.Ledger.Contexts qualified as Contexts
+import Plutus.V1.Ledger.Contexts.Extra qualified as Contexts.Extra
+import Plutus.V1.Ledger.Scripts (
+  Validator,
+  mkValidatorScript,
+ )
+import Plutus.V1.Ledger.Value qualified as Value
+import PlutusTx (BuiltinData, applyCode)
+import PlutusTx.Prelude (fmap, fromMaybe, trace, traceIfFalse, uniqueElement, zero, ($), (&&), (-), (==))
+import PlutusTx.TH (compile)
+import Wallet.Emulator.Types (WalletNumber (WalletNumber))
+import Wallet.Emulator.Wallet (Wallet, fromWalletNumber, walletPubKeyHash)
+
+--------------------------------------------------------------------------------
+
+main :: IO ()
+main = defaultMain tests
+
+tests :: TestTree
+tests =
+  localOption (QuickCheckTests 100) $
```

Comment:
`make test` ran 800 tests for "Output secret" property. Is the option passed properly?

Reply:
We use coverage in our testing. And it is little tricky.

What does it do? We have some `tolerance` and `certainty` for coverage.

Every `powerOfTwo * 100` tests it checks (`kLow` and `kHigh` - coefficients depending on the `certainty`):
if one of them is True - we can stop testing:
```
Successful coverage:
willson numOfClasified numOfSuccessed kLow >= tolerance * desiredCoverage

Unsuccessful coverage:
willson numOfClasified numOfSuccessed kHight < desiredCoverage
```
else - continue the testing

[About Willson law](https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Wilson_score_interval).

As far as I understand, until a coverage decision is made, we cannot stop testing and QuickCheckTests will not work.

Reply:
Ok, got it.